trainer/trainer.py

```diff
@@ -1014,12 +1014,16 @@ def _optimize(
                         loss_dict["amp_scaler"] = scaler.get_scale()  # for logging
                     update_lr_scheduler = scale_prev <= scaler.get_scale()
         else:
+            self.callbacks.before_main_optimizer_step(self, loss_dict, optimizer)
```

Comment:
I'd call this "before_backward_pass" 

and move ```self.model.apply_gradient_clipping``` in ```before_gradient_clipping```

Reply:
@erogol  do you mean wrapping `self.model.apply_gradient_clipping` into a separate callback?